src/Features/Core/Portable/Diagnostics/AnalyzerHelper.cs

```diff
@@ -435,81 +381,64 @@ public static bool SupportAnalysisKind(this DiagnosticAnalyzerService service, D
                     _ => throw ExceptionUtilities.UnexpectedValue(kind),
                 };
 
-                var diagnostics = (await analyzeAsync.ConfigureAwait(false)).NullToEmpty();
-                if (compilation != null)
-                {
-                    return CompilationWithAnalyzers.GetEffectiveDiagnostics(diagnostics, compilation);
-                }
+                diagnostics = (await analyzeAsync.ConfigureAwait(false)).NullToEmpty();
 
 #if DEBUG
                 // since all DocumentDiagnosticAnalyzers are from internal users, we only do debug check. also this can be expensive at runtime
                 // since it requires await. if we find any offender through NFW, we should be able to fix those since all those should
                 // from intern teams.
                 await VerifyDiagnosticLocationsAsync(diagnostics, document.Project, cancellationToken).ConfigureAwait(false);
 #endif
-
-                return diagnostics;
             }
             catch (Exception e) when (!IsCanceled(e, cancellationToken))
             {
-                OnAnalyzerException(service, analyzer, document.Project.Id, compilation, e, logAggregator);
-                return ImmutableArray<Diagnostic>.Empty;
+                diagnostics = ImmutableArray.Create(CreateAnalyzerExceptionDiagnostic(analyzer, e));
```

Comment:
Ah, we are explicitly reporting analyzer exception diagnostics for internal analyzers here. That resolves my original concern.